op-challenger/game/keccak/matrix/matrix.go

```diff
@@ -17,10 +18,61 @@ type StateMatrix struct {
 }
 
 var (
-	ErrInvalidMaxLen = errors.New("invalid max length to absorb")
-	uint256Size      = 32
+	ErrInvalidMaxLen            = errors.New("invalid max length to absorb")
+	ErrIncorrectCommitmentCount = errors.New("incorrect number of commitments for input length")
+	uint256Size                 = 32
 )
 
+// VerifyPreimage checks the claimed state commitments are correct for the provided preimage data.
+// Returns nil if the preimage is valid, otherwise returns a [types.Challenge] proving the fault.
+func VerifyPreimage(data io.Reader, commitments []common.Hash) (*types.Challenge, error) {
+	s := NewStateMatrix()
+	m := s.PackState()
+	var prestate types.Leaf
+	for i := 0; ; i++ {
+		unpaddedLeaf, err := s.absorbNextLeafInput(data)
+		if err != nil && !errors.Is(err, io.EOF) {
+			return nil, fmt.Errorf("failed to verify inputs: %w", err)
+		}
+		isEOF := errors.Is(err, io.EOF)
+		validCommitment := s.StateCommitment()
+		if i >= len(commitments) {
+			// There should have been more commitments.
+			// The contracts should prevent this so it can't be challenged, return an error
+			return nil, ErrIncorrectCommitmentCount
+		}
+		claimedCommitment := commitments[i]
+
+		var paddedLeaf [types.BlockSize]byte
+		copy(paddedLeaf[:], unpaddedLeaf)
+		// TODO(client-pod#480): Add actual keccak padding to ensure the merkle proofs are correct
+		poststate := types.Leaf{
+			Input:           paddedLeaf,
+			Index:           big.NewInt(int64(i)),
+			StateCommitment: claimedCommitment,
+		}
+
+		if validCommitment != claimedCommitment {
+			return &types.Challenge{
+				StateMatrix: m,
+				Prestate:    prestate,
+				Poststate:   poststate,
+			}, nil
+		}
+		if isEOF {
+			if i < len(commitments)-1 {
+				// We got too many commitments
+				// The contracts should prevent this so it can't be challenged, return an error
```

Comment:
Is this a correct statement?
```suggestion
// The contracts should prevent too many commitments so it can't be challenged
```

Reply:
Yep: https://github.com/ethereum-optimism/optimism/blob/develop/packages/contracts-bedrock/src/cannon/PreimageOracle.sol#L325

(And that's a quote from Clabby because I asked him the same thing 😆 ).  Most importantly, there's no way to challenge a preimage that has the wrong number of state commitments. Particularly for too few commitments, you'd need to post a merkle proof that the missing commitment _was_ included but then it can't be missing.